service/matching/matching_engine.go

```diff
@@ -837,6 +837,118 @@ func (e *matchingEngineImpl) listTaskQueuePartitions(request *matchingservice.Li
+	if err != nil {
+		return nil, err
+	}
+
+	// todo carly: do we need to update updateOptions.TaskQueueLimitPerBuildId or updateOptions.KnownVersion like in UpdateWorkerBuildIdCompatibility?
+	updateOptions := UserDataUpdateOptions{}
+
+	err = tqMgr.UpdateUserData(ctx, updateOptions, func(data *persistencespb.TaskQueueUserData) (*persistencespb.TaskQueueUserData, bool, error) {
+		clk := data.GetClock()
+		if clk == nil {
+			tmp := hlc.Zero(e.clusterMeta.GetClusterID())
+			clk = tmp
+		}
+		updatedClock := hlc.Next(clk, e.timeSource)
+		var versioningData *persistencespb.VersioningData
+		switch req.GetOperation().(type) {
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_InsertAssignmentRule:
+			versioningData, err = InsertAssignmentRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetInsertAssignmentRule(),
+				e.config.VersionAssignmentRuleLimitPerQueue(ns.Name().String()),
+			)
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_ReplaceAssignmentRule:
+			versioningData, err = ReplaceAssignmentRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetReplaceAssignmentRule(),
+			)
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_DeleteAssignmentRule:
+			versioningData, err = DeleteAssignmentRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetDeleteAssignmentRule(),
+			)
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_InsertCompatibleRedirectRule:
+			versioningData, err = InsertCompatibleRedirectRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetInsertCompatibleRedirectRule(),
+				e.config.VersionRedirectRuleLimitPerQueue(ns.Name().String()),
+			)
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_ReplaceCompatibleRedirectRule:
+			versioningData, err = ReplaceCompatibleRedirectRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetReplaceCompatibleRedirectRule(),
+			)
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_DeleteCompatibleRedirectRule:
+			versioningData, err = DeleteCompatibleRedirectRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetDeleteCompatibleRedirectRule(),
+			)
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_CommitBuildId_:
+			// todo carly: later PR
+		}
+		if err != nil {
+			// operation can't be completed due to limits. no action, do not replicate, report error
+			return nil, false, err
+		}
+		// No need to keep tombstones around for replication
+		if ns.ReplicationPolicy() != namespace.ReplicationPolicyMultiCluster {
+			versioningData = CleanupRuleTombstones(versioningData, e.config.VersionDeletedRuleRetentionTime(ns.Name().String()))
+		}
+		// Avoid mutation
+		ret := common.CloneProto(data)
+		ret.Clock = updatedClock
+		ret.VersioningData = versioningData
+		return ret, true, nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Only clear tombstones after they have been replicated. This is only really an issue if
```

Comment:
this is right. But I think we can rely on the retention time being long enough because it's in our control. We can add a note for that config saying that a very small retention could cause problems for replication. 
Assuming the retention time is long enough, we can cleanup before retention and get rid of this secondary update.